src/ucp/core/ucp_am.c

```diff
@@ -0,0 +1,468 @@
+    }
+    ucp_recv_desc_release(rdesc);
+}
+
+ucs_status_t ucp_worker_set_am_handler(ucp_worker_h worker, uint16_t id, 
+                                       ucp_am_callback_t cb, void *arg, 
+                                       uint32_t flags)
+{
+    if (id >= worker->am_cb_array_len) {
+        size_t num_entries = id + (id % AM_BLOCK);
+        worker->am_cbs = ucs_realloc(worker->am_cbs, num_entries * 
+                                     sizeof(ucp_worker_am_entry_t),
+                                     "UCP AM callback array");
+        
+        memset(worker->am_cbs + worker->am_cb_array_len, 
+               0, (num_entries - worker->am_cb_array_len)
+               * sizeof(ucp_worker_am_entry_t));
+        
+        worker->am_cb_array_len = num_entries;
+    }
+    worker->am_cbs[id].cb = cb;
+    worker->am_cbs[id].context = arg;
+    worker->am_cbs[id].flags = flags;
+
+    return UCS_OK;
+}
+
+size_t bcopy_pack_args_single(void *dest, void *arg)
+{
+    ucp_am_hdr_t *hdr = dest;
+    ucp_request_t *req = arg;
+    size_t length;
+
+    ucs_assert(req->send.state.dt.offset == 0);
+
+    hdr->am_id = req->send.am.am_id;
+    hdr->total_size = req->send.length;
+
+    length = ucp_dt_pack(req->send.ep->worker, req->send.datatype,
+                         UCT_MD_MEM_TYPE_HOST, hdr + 1, req->send.buffer,
+                         &req->send.state.dt, req->send.length);
+    ucs_assert(length == req->send.length);
+
+    return sizeof(ucp_am_hdr_t) + length;
+}
+
+size_t bcopy_pack_args_first(void *dest, void *arg){
+    ucp_am_long_hdr_t *hdr = dest;
+    ucp_request_t *req = arg;
+    size_t length;
+
+    length = ucp_ep_get_max_bcopy(req->send.ep, req->send.lane) -
+                                  sizeof(ucp_am_long_hdr_t);
+    hdr->total_size = req->send.length;
+    hdr->am_id      = req->send.am.am_id;
+    hdr->msg_id         = req->send.tag.message_id;
+    hdr->ep             = ucp_request_get_dest_ep_ptr(req);
+    hdr->offset         = req->send.state.dt.offset;
+    return sizeof(*hdr) + ucp_dt_pack(req->send.ep->worker, 
+                                      req->send.datatype, 
+                                      UCT_MD_MEM_TYPE_HOST,
+                                      hdr + 1, req->send.buffer,
+                                      &req->send.state.dt, length);
+                                                    
+}
+
+size_t bcopy_pack_args_mid(void *dest, void *arg)
+{
+    ucp_am_long_hdr_t *hdr = dest;
+    ucp_request_t *req = arg;
+    size_t length;
+    length = ucs_min(ucp_ep_get_max_bcopy(req->send.ep, req->send.lane) - 
+                     sizeof(*hdr),
+                     req->send.length - req->send.state.dt.offset);
+
+    hdr->msg_id         = req->send.tag.message_id;
+    hdr->offset         = req->send.state.dt.offset;
+    hdr->ep             = ucp_request_get_dest_ep_ptr(req);
+    hdr->am_id      = req->send.am.am_id;
+    hdr->total_size = req->send.length;
```

Comment:
Because we do not have a guarantee of ordering, the first one to arrive allocated the buffer for all uct active messages to complete the ucp active message. Therefore everyone needs to know how large a buffer to allocate. Originally if the first uct am did not arrive first, we would store the uct am in a list until the first message arrived and then the out-of-order ams would be copied into the buffer. However, we found it difficult to test if this was implemented correctly, so I opted to just have the first message to arrive allocate the long buffer and then everyone copies themselves in as they arrive.

Reply:
As I see this implementation doesn't use multi-lane, Single UCT lane should guarantee ordering.
Since multi-fragment protocols are not latency critical, i think need to remember that out of ordering is possible on protocol layer when multi-lane will be enabled, please document it somewhere in code.